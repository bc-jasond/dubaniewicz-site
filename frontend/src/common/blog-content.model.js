import React from 'react';
import styled from 'styled-components';
import {
  NODE_TYPE_CODE,
  NODE_TYPE_ITALIC,
  NODE_TYPE_LI,
  NODE_TYPE_LINK,
  NODE_TYPE_OL,
  NODE_TYPE_P,
  NODE_TYPE_A,
  NODE_TYPE_SECTION_CODE,
  NODE_TYPE_SECTION_CONTENT,
  NODE_TYPE_SECTION_H1,
  NODE_TYPE_SECTION_H2,
  NODE_TYPE_SECTION_IMAGE,
  NODE_TYPE_SECTION_QUOTE,
  NODE_TYPE_SECTION_SPACER,
  NODE_TYPE_SECTION_POSTLINK,
  NODE_TYPE_SITEINFO,
  NODE_TYPE_STRIKE,
  NODE_TYPE_TEXT,
  NODE_TYPE_ROOT,
} from './constants';
import {
  H1,
  H2,
  ContentSection,
  SpacerSection,
  CodeSection,
  P,
  Pre,
  A,
  Ol,
  Li,
  LinkStyled,
  SiteInfo,
  StrikeText,
  ItalicText,
  Code,
  Figure,
  ImagePlaceholderContainer,
  ImagePlaceholderFill,
  Img,
  FigureCaption,
  ImageSection,
} from './shared-styled-components';

function s4() {
  return Math.floor((1 + Math.random()) * 0x10000)
    .toString(16)
    .substring(1);
}

// TODO: this is a placeholder to be able to set the caret in an empty tag

export class BlogPostNode {
  constructor({ type, id, parent_id, post_id, content, meta }, parent = null) {
    this.type = type;
    this.id = id || s4();
    this.parent_id = parent_id;
    this.post_id = post_id;
    this.content = content;
    this.meta = meta;
    
    this.parent = parent;
    this.childNodes = [];
  }
  
  canHaveChildren() {
    return ![
      NODE_TYPE_CODE,
      NODE_TYPE_SECTION_CODE,
      NODE_TYPE_SECTION_IMAGE,
      NODE_TYPE_SECTION_QUOTE,
      NODE_TYPE_SECTION_SPACER,
      NODE_TYPE_SECTION_POSTLINK,
      NODE_TYPE_TEXT,
    ].includes(this.type)
  }
  
  getNextSibling() {
    if (!this.parent
      || this.parent.childNodes.indexOf(this) === this.parent.childNodes.size - 1) {
      return null;
    }
    const idx = this.parent.childNodes.indexOf(this);
    return this.parent.childNodes.get(idx + 1);
  }
  
  deleteChildNode(node) {
    const idx = this.childNodes.indexOf(node);
    if (idx > -1) {
      this.childNodes.delete(idx);
      return true;
    }
    return false;
  }
  
  findById(id) {
    const queue = [this];
    while (queue.length) {
      const current = queue.pop();
      if (current.id === id) {
        return current;
      }
      if (current.childNodes.size) {
        queue.unshift(...current.childNodes.toArray())
      }
    }
    return null;
  }
  
  getKey() {
  
  }
  
  updateContent(content) {
    let textNode;
    if (this.childNodes.size === 0) {
      textNode = new BlogPostNode({ type: NODE_TYPE_TEXT }, this);
    } else {
      // TODO: fix this - assume there's only one child and it's a text node
      textNode = this.childNodes.get(0);
      this.childNodes.delete(0);
    }
    textNode.content = content;
    this.childNodes = this.childNodes.push(textNode);
  }
  
  toJSON() {
    const raw = { ...this };
    delete (raw.parent); // prevent circular reference issue
    // raw.childNodes = raw.childNodes.map(child => child.toJson());
    return raw;
  }
  
  render() {
    switch (this.type) {
      /**
       * SPECIAL TYPES
       */
      case NODE_TYPE_ROOT:
        return (
          <div data-type="root" name={this.id}>
            {this.childNodes.map(node => node.render())}
          </div>
        );
      case NODE_TYPE_TEXT:
        return this.content;
      case NODE_TYPE_LI:
        return (<Li>{this.childNodes.map(node => node.render())}</Li>);
      /**
       * SECTIONS
       */
      case NODE_TYPE_SECTION_SPACER:
        return (<SpacerSection />);
      case NODE_TYPE_SECTION_H1:
        return (
          <H1 data-type="h1" name={this.id}>
            {this.childNodes.map(node => node.render())}
          </H1>
        );
      case NODE_TYPE_SECTION_H2:
        return (<H2>{this.childNodes.map(node => node.render())}</H2>);
      case NODE_TYPE_SECTION_CONTENT:
        return (<ContentSection data-type="content"
                                name={this.id}>{this.childNodes.map(node => node.render())}</ContentSection>);
      case NODE_TYPE_SECTION_CODE:
        const { lines } = this.meta;
        return (
          <CodeSection>
            {lines.map((line, idx) => (<Pre key={idx}>{line}</Pre>))}
          </CodeSection>
        );
      case NODE_TYPE_SECTION_IMAGE: {
        const { width, height, url, caption } = this.meta;
        return (
          <ImageSection>
            <Figure>
              <ImagePlaceholderContainer w={width} h={height}>
                <ImagePlaceholderFill w={width} h={height} />
                <Img src={url} />
              </ImagePlaceholderContainer>
              <FigureCaption>{caption}</FigureCaption>
            </Figure>
          </ImageSection>
        );
      }
      case NODE_TYPE_SECTION_QUOTE: {
        const { quote, url, author, context } = this.meta;
        return (
          <ContentSection>
            <P>ðŸ’¡Remember: <ItalicText>{quote}<A href={url}>{author}</A>{context}
            </ItalicText></P>
          </ContentSection>
        );
      }
      case NODE_TYPE_SECTION_POSTLINK: {
        const { to } = this.meta;
        const Centered = styled.div`
          text-align: center;
        `;
        const PLarger = styled(P)`
          font-size: larger;
        `;
        return (
          <Centered>
            <ContentSection>
              <PLarger><SiteInfo>Thanks for reading</SiteInfo></PLarger>
            </ContentSection>
            {to && (
              <H2>
                Next Post: <LinkStyled to={to}>{this.content}</LinkStyled>
              </H2>
            )}
            <H2>
              ðŸ‘ˆ <LinkStyled to="/posts">Back to all Posts</LinkStyled>
            </H2>
          </Centered>
        );
      }
      /**
       * FORMATTING TYPES
       */
      case NODE_TYPE_CODE:
        return (<Code>{this.content}</Code>);
      case NODE_TYPE_P:
        return (
          <P data-type="p" name={this.id}>
            {this.childNodes.map(node => node.render())}
          </P>
        );
      case NODE_TYPE_OL:
        return (<Ol>{this.childNodes.map(node => node.render())}</Ol>);
      case NODE_TYPE_LINK:
        return (<LinkStyled to={this.content}>{this.childNodes.map(node => node.render())}</LinkStyled>);
      case NODE_TYPE_A:
        return (<A href={this.content}>{this.childNodes.map(node => node.render())}</A>);
      case NODE_TYPE_SITEINFO:
        return (<SiteInfo>{this.childNodes.map(node => node.render())}</SiteInfo>);
      case NODE_TYPE_STRIKE:
        return (<StrikeText>{this.childNodes.map(node => node.render())}</StrikeText>);
      case NODE_TYPE_ITALIC:
        return (<ItalicText>{this.childNodes.map(node => node.render())}</ItalicText>);
      
    }
  }
}

/**
 * Build up the blog post content DOM like model using BFS
 *
 * @param nodes - a hash of all content node data keyed off of parent_id
 * @returns BlogPostNode - the root node
 */
export function getContentTree(nodes) {
  const root = new BlogPostNode(nodes['null'][0]);
  let current;
  const queue = [root];
  
  while (queue.length) {
    current = queue.shift();
    const children = nodes[current.id] || [];
    children.forEach(node => {
      const childNode = new BlogPostNode(node, current);
      current.childNodes.push(childNode);
      queue.push(childNode);
    })
  }
  
  return root;
}

// opinionated section nodes - can't have children
// these could be interesting in the editor 'add' menu but, they're currently supported with existing node types
// export const Gotcha = {} // facepalm emoji
// export const Tangent = {} // horse emoji
// export const Shrug = {} // shrug emoji

