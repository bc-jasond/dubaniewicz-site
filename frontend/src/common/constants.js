export const NODE_TYPE_ROOT = 'root';
// sections for layout - can only be children of the root node 'blogContent'
export const NODE_TYPE_SECTION_H1 = 'h1';
export const NODE_TYPE_SECTION_H2 = 'h2';
export const NODE_TYPE_SECTION_SPACER = 'spacer';
export const NODE_TYPE_SECTION_CONTENT = 'content';
// opinionated sections - have fixed format, can't have children
export const NODE_TYPE_SECTION_CODE = 'codesection';
export const NODE_TYPE_SECTION_IMAGE = 'image';
export const NODE_TYPE_SECTION_QUOTE = 'quote';
export const NODE_TYPE_SECTION_POSTLINK = 'postlink';

// nodes for content - must have a parent node of a section type
export const NODE_TYPE_TEXT = 'text';  // the #text or DOM nodeType 3
export const NODE_TYPE_P = 'p';
export const NODE_TYPE_PRE = 'pre';
export const NODE_TYPE_OL = 'ol';
export const NODE_TYPE_LI = 'li';
export const NODE_TYPE_A = 'a';
export const NODE_TYPE_LINK = 'link';
export const NODE_TYPE_CODE = 'code';
export const NODE_TYPE_SITEINFO = 'siteinfo';
export const NODE_TYPE_ITALIC = 'italic';
export const NODE_TYPE_STRIKE = 'strike';

// EDITOR
// TODO: this is a placeholder to be able to set the caret in an empty tag
export const NEW_POST_URL_ID = 'new';
export const ZERO_LENGTH_CHAR = '\u200B';
export const ENTER_KEY = 13;
export const BACKSPACE_KEY = 8;
export const UP_ARROW = 38;
export const DOWN_ARROW = 40;
export const LEFT_ARROW = 37;
export const RIGHT_ARROW = 39;

// ENV
export const API_URL = 'http://localhost:3001';

// AUTH
export const AUTH_TOKEN_KEY = 'dubaniewicz-token';
export const SESSION_KEY = 'dubaniewicz-session';