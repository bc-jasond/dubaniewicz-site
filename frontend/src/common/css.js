export const lightGrey = 'rgba(0,0,0,.05)';
export const grey = 'rgba(0,0,0,.54)';
export const darkGrey = 'rgba(0,0,0,.84)';
export const lightBlue = '#A0CFE1';