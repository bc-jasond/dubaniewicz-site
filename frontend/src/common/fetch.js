import { API_URL } from './constants';
import { getAuthToken } from './session';

function getBaseConfig() {
  return {
    // Default options are marked with *
    mode: 'cors', // no-cors, cors, *same-origin
    cache: 'no-cache', // *default, no-cache, reload, force-cache, only-if-cached
    headers: {
      'Content-Type': 'application/json',
      Authorization: getAuthToken(),
    },
    redirect: 'follow', // manual, *follow, error
    referrer: 'no-referrer', // no-referrer, *client
  };
}

export async function apiGet(url) {
  try {
    Pace.start();
    const config = getBaseConfig();
    config.method = 'GET';
    const response = await fetch(`${API_URL}${url}`, config);
    Pace.stop();
    return response.json();
  } catch (err) {
    Pace.stop();
    console.error('Fetch GET Error: ', err);
    throw err;
  }
}

export async function apiPost(url, data) {
  try {
    Pace.start();
    const config = getBaseConfig();
    config.method = 'POST';
    config.body = JSON.stringify(data); // body data type must match "Content-Type" header
    const response = await fetch(`${API_URL}${url}`, config);
    if (response.status.toString(10).charAt(0) !== '2') {
      throw new Error(`${response.status} - ${response.statusText}`);
    }
    Pace.stop();
    return response.json();
  } catch (err) {
    Pace.stop();
    console.error('Fetch POST Error: ', err);
    throw err;
  }
}

export async function apiDelete(url) {
  try {
    Pace.start();
    const config = getBaseConfig();
    config.method = 'DELETE';
    const response = await fetch(`${API_URL}${url}`, config);
    Pace.stop();
    return response.json();
  } catch (err) {
    Pace.stop();
    console.error('Fetch DELETE Error: ', err);
    throw err;
  }
}
