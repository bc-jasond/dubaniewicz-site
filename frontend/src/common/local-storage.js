import { fromJS, Map } from 'immutable';
import { reviver } from '../pages/edit/document-model';
import { FILBERT_LOCALSTORAGE_KEY } from './constants';

function getLocalStorageContents() {
  let session;
  try {
    session = JSON.parse(localStorage.getItem(FILBERT_LOCALSTORAGE_KEY) || {});
    return fromJS(session, reviver);
  } catch (err) {
    return Map();
  }
}

export function get(key, defaultValue) {
  return getLocalStorageContents().get(key, defaultValue);
}

export function set(key, value) {
  localStorage.setItem(
    FILBERT_LOCALSTORAGE_KEY,
    JSON.stringify(
      getLocalStorageContents()
        .set(key, value)
        .toJS()
    )
  );
}
