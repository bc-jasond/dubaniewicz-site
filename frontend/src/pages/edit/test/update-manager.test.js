import { Map, List } from 'immutable';

import UpdateManager, { characterDiffSize } from '../update-manager';
import DocumentModel from '../document-model';
import {
  overrideConsole,
  mockLocalStorage
} from '../../../common/test-helpers';
import { clearForTests } from '../../../common/local-storage';
import {
  formattedPId,
  h2Id,
  imgId,
  pre2Id,
  testPostWithAllTypesJS
} from './test-post-with-all-types';
import {
  HISTORY_KEY_NODE_UPDATES,
  HISTORY_KEY_UNDO,
  HISTORY_KEY_REDO,
  HISTORY_KEY_UNDO_OFFSETS,
  HISTORY_KEY_UNDO_UPDATES,
  HISTORY_KEY_REDO_OFFSETS,
  HISTORY_KEY_REDO_UPDATES,
  NODE_TYPE_H1
} from '../../../common/constants';
import * as api from '../../../common/fetch';

jest.mock('../../../common/fetch', () => ({
  __esModule: true,
  apiPost: jest.fn(async () => ({}))
}));
jest.useFakeTimers();

const { post, contentNodes } = testPostWithAllTypesJS;

overrideConsole();
mockLocalStorage();
let updateManager;

beforeEach(() => {
  clearForTests();
  localStorage.clear();
  updateManager = new UpdateManager();
  updateManager.init(post);
});

describe('UpdateManager', () => {
  test('init method', () => {
    expect(updateManager).toMatchSnapshot();
  });
  test('stageNodeUpdate method', () => {
    updateManager.stageNodeUpdate();
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    updateManager.stageNodeUpdate(null);
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    updateManager.stageNodeUpdate(Map({ id: 'null' }));
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    updateManager.stageNodeUpdate(Map({ id: 'undefined' }));
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    // test that last-write-wins, update overwrites delete
    updateManager.stageNodeDelete(Map({ id: '1234' }));
    updateManager.stageNodeUpdate(Map({ id: '1234' }));
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(1);
    expect(updateManager).toMatchSnapshot();
  });
  test('stageNodeDelete method', () => {
    updateManager.stageNodeDelete();
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    updateManager.stageNodeDelete(null);
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    updateManager.stageNodeDelete(Map({ id: 'null' }));
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    updateManager.stageNodeDelete(Map({ id: 'undefined' }));
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    // test that last-write-wins, delete overwrites update
    updateManager.stageNodeUpdate(Map({ id: '1234' }));
    updateManager.stageNodeDelete(Map({ id: '1234' }));
    expect(updateManager).toMatchSnapshot();
  });
  test('addPostIdToUpdates method', () => {
    // mimic a "not-yet-saved" post
    updateManager.init({});
    updateManager.stageNodeUpdate(Map({ id: 'abcd' }));
    expect(
      updateManager[HISTORY_KEY_NODE_UPDATES].get('abcd').get('post_id')
    ).toBeUndefined();
    updateManager.addPostIdToUpdates(1);
    expect(
      updateManager[HISTORY_KEY_NODE_UPDATES].get('abcd').get('post_id')
    ).toBe(1);
  });
  test('clearUpdates method', () => {
    updateManager.stageNodeUpdate(Map({ id: '1111' }));
    updateManager.stageNodeUpdate(Map({ id: 'abcd' }));
    updateManager.stageNodeDelete(Map({ id: '2222' }));
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(3);
    updateManager.clearUpdates();
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    // can call clearUpdates on empty
    updateManager.clearUpdates();
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
  });
  /**
   * INTEGRATION TESTS - THESE DEPEND ON DocumentModel, or you have to simulate DocumentModel behavior which seems of dubious value
   */
  test('addToUndoHistory - delete a node in the middle', () => {
    const prevSelectionOffsets = { is: 'previousOffsets' };
    const selectionOffsets = { is: 'currentOffsets' };
    const doc = new DocumentModel();
    const firstNode = doc.init(post, updateManager, contentNodes);
    let prevNodesById = doc.nodesById;
    // create history entry when deleting a node
    doc.delete(doc.getNode(imgId));
    // 1 for the deleted node, 1 for the prev node to update next_sibling_id reference
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(2);
    expect(updateManager[HISTORY_KEY_UNDO].size).toBe(0);
    expect(updateManager[HISTORY_KEY_NODE_UPDATES]).toMatchSnapshot();
    updateManager.addToUndoHistory(
      prevNodesById,
      prevSelectionOffsets,
      selectionOffsets
    );
    // there should be 1 history entry in the undo stack
    expect(updateManager[HISTORY_KEY_UNDO].size).toBe(1);
    const firstHistory = updateManager[HISTORY_KEY_UNDO].get(0);
    // there should be 2 nodes in the undo
    expect(firstHistory.get(HISTORY_KEY_UNDO_UPDATES).size).toBe(2);
    expect(firstHistory.get(HISTORY_KEY_UNDO_UPDATES)).toMatchSnapshot();
    expect(firstHistory.get(HISTORY_KEY_UNDO_OFFSETS)).toBe(
      prevSelectionOffsets
    );
    // and 2 in the redo - they should match the current updates
    expect(firstHistory.get(HISTORY_KEY_REDO_UPDATES).size).toBe(2);
    expect(firstHistory.get(HISTORY_KEY_REDO_UPDATES)).toEqual(
      updateManager[HISTORY_KEY_NODE_UPDATES]
    );
    expect(firstHistory.get(HISTORY_KEY_REDO_OFFSETS)).toBe(selectionOffsets);
  });
  test('addToUndoHistory - change a node`s type from P -> H1', () => {
    const prevSelectionOffsets = { is: 'previousOffsets' };
    const selectionOffsets = { is: 'currentOffsets' };
    const doc = new DocumentModel();
    const firstNode = doc.init(post, updateManager, contentNodes);
    let prevNodesById = doc.nodesById;
    // create history entry when changing a node's type
    doc.update(
      doc
        .getNode(formattedPId)
        .set('type', NODE_TYPE_H1)
        .set('meta', Map())
    );
    // 1 for the deleted node, 1 for the prev node to update next_sibling_id reference
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(1);
    expect(updateManager[HISTORY_KEY_UNDO].size).toBe(0);
    expect(updateManager[HISTORY_KEY_NODE_UPDATES]).toMatchSnapshot();
    updateManager.addToUndoHistory(
      prevNodesById,
      prevSelectionOffsets,
      selectionOffsets
    );
    // there should be 1 history entry in the undo stack
    expect(updateManager[HISTORY_KEY_UNDO].size).toBe(1);
    const firstHistory = updateManager[HISTORY_KEY_UNDO].get(0);
    // there should be 2 nodes in the undo
    expect(firstHistory.get(HISTORY_KEY_UNDO_UPDATES).size).toBe(1);
    expect(firstHistory.get(HISTORY_KEY_UNDO_UPDATES)).toMatchSnapshot();
    expect(firstHistory.get(HISTORY_KEY_UNDO_OFFSETS)).toBe(
      prevSelectionOffsets
    );
    // and 2 in the redo - they should match the current updates
    expect(firstHistory.get(HISTORY_KEY_REDO_UPDATES).size).toBe(1);
    expect(firstHistory.get(HISTORY_KEY_REDO_UPDATES)).toEqual(
      updateManager[HISTORY_KEY_NODE_UPDATES]
    );
    expect(firstHistory.get(HISTORY_KEY_REDO_OFFSETS)).toBe(selectionOffsets);
  });
  test('addToUndoHistory - simulate typing inside content field of one node', () => {
    const prevSelectionOffsets = { is: 'previousOffsets' };
    const selectionOffsets = { is: 'currentOffsets' };
    const doc = new DocumentModel();
    const firstNode = doc.init(post, updateManager, contentNodes);
    let prevNodesById = doc.nodesById;
    const newContent = 'Sammys!';
    // loop over the string, adding one char at a time to simulate typing
    for (let i = 0; i < newContent.length - 1; i++) {
      doc.update(
        doc
          .getNode(h2Id)
          .set(
            'content',
            `${doc.getNode(h2Id).get('content')}${newContent.charAt(i)}`
          )
      );
      // since updates are keyed off of nodeId, there will only be 1 update
      expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(1);
      expect(updateManager[HISTORY_KEY_UNDO].size).toBe(0);
      updateManager.addToUndoHistory(
        prevNodesById,
        i === 0 ? prevSelectionOffsets : { is: `wrongOffsets${i}` },
        selectionOffsets
      );
      // there shouldn't a history entry until we reach the change threshold
      expect(updateManager[HISTORY_KEY_UNDO].size).toBe(0);
    }
    doc.update(
      doc
        .getNode(h2Id)
        .set(
          'content',
          `${doc.getNode(h2Id).get('content')}${newContent.charAt(
            newContent.length - 1
          )}`
        )
    );
    // since updates are keyed off of nodeId, there will only be 1 update
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(1);
    expect(updateManager[HISTORY_KEY_NODE_UPDATES]).toMatchSnapshot();
    expect(updateManager[HISTORY_KEY_UNDO].size).toBe(0);
    updateManager.addToUndoHistory(
      prevNodesById,
      { is: `wrongOffsets${newContent.length - 1}` },
      selectionOffsets
    );
    // we should have hit the threshold now
    expect(updateManager[HISTORY_KEY_UNDO].size).toBe(1);
    const firstHistory = updateManager[HISTORY_KEY_UNDO].get(0);
    expect(firstHistory.get(HISTORY_KEY_UNDO_UPDATES).size).toBe(1);
    expect(firstHistory.get(HISTORY_KEY_UNDO_UPDATES)).toMatchSnapshot();
    expect(firstHistory.get(HISTORY_KEY_UNDO_OFFSETS)).toBe(
      prevSelectionOffsets
    );
    expect(firstHistory.get(HISTORY_KEY_REDO_UPDATES).size).toBe(1);
    expect(firstHistory.get(HISTORY_KEY_REDO_UPDATES)).toEqual(
      updateManager[HISTORY_KEY_NODE_UPDATES]
    );
    expect(firstHistory.get(HISTORY_KEY_REDO_OFFSETS)).toBe(selectionOffsets);
  });
  test.todo(
    'addToUndoHistory - simulate select and type inside content field of one node'
  );
  test.todo('addToUndoHistory - simulate select and type across nodes');
  test('applyUpdates', () => {
    const doc = new DocumentModel();
    const firstNode = doc.init(post, updateManager, contentNodes);
    const prevNodesById = doc.nodesById;
    const h2 = doc.getNode(h2Id);
    const img = doc.getNode(imgId);
    doc.update(h2.set('content', 'foo'));
    doc.delete(img);
    const updates = updateManager[HISTORY_KEY_NODE_UPDATES];
    updateManager.clearUpdates();
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
    const updatedNodes = updateManager.applyUpdates(updates, prevNodesById);
    expect(updateManager[HISTORY_KEY_NODE_UPDATES]).toEqual(updates);
    expect(updatedNodes.get(h2Id).get('content')).toBe('foo');
    expect(updatedNodes.get(imgId)).toBeFalsy();
  });
  test.todo('undo');
  test.todo('redo');
  test('getKey', () => {
    expect(updateManager.getKey('foo')).toMatchInlineSnapshot(`"175-foo"`);
  });
  test('getPostIdNamespaceValue & setPostIdNamespaceValue', () => {
    updateManager.setPostIdNamespaceValue('foo', 'bar');
    expect(updateManager.getPostIdNamespaceValue('foo', 'qux')).toBe('bar');
  });
  test('nodeUpdates, redoHistory, undoHistory', () => {
    expect(updateManager[HISTORY_KEY_NODE_UPDATES]).toBe(Map());
    expect(updateManager[HISTORY_KEY_REDO]).toBe(List());
    expect(updateManager[HISTORY_KEY_UNDO]).toBe(List());
  });
  test('saveContentBatch', async () => {
    // empty updates should not make API call
    await updateManager.saveContentBatch();
    expect(api.apiPost).not.toHaveBeenCalled();
    // stage some updates
    updateManager.stageNodeUpdate(Map({ id: '1111' }));
    updateManager.stageNodeUpdate(Map({ id: 'abcd' }));
    updateManager.stageNodeDelete(Map({ id: '2222' }));
    const entries = Object.entries(
      updateManager[HISTORY_KEY_NODE_UPDATES].toJS()
    );
    // test api failure - updates should remain
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(3);
    api.apiPost = jest.fn(async () => ({ error: true }));
    await updateManager.saveContentBatch();
    expect(api.apiPost).toHaveBeenCalledWith('/content', entries);
    // test api success - updates are cleared
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(3);
    api.apiPost = jest.fn(async () => ({}));
    await updateManager.saveContentBatch();
    expect(api.apiPost).toHaveBeenCalledWith('/content', entries);
    expect(updateManager[HISTORY_KEY_NODE_UPDATES].size).toBe(0);
  });
  test('saveContentDebounce', () => {
    updateManager.saveContentBatch = jest.fn();
    updateManager.saveContentBatchDebounce();
    updateManager.saveContentBatchDebounce();
    updateManager.saveContentBatchDebounce();
    jest.runAllTimers();
    expect(updateManager.saveContentBatch).toHaveBeenCalledTimes(1);
  });
});
