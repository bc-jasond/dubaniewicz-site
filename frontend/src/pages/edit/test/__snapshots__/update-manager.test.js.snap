// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateManager init method 1`] = `
UpdateManager {
  "commitTimeoutId": undefined,
  "lastUndoHistoryPush": undefined,
  "post": Immutable.Map {
    "abstract": "Creating a simple React ( + Babel + Webpack + yarn) starter project is actually proving to be difficult starting from the documentation.",
    "created": "2019-05-18 17:28:55",
    "published": "2019-01-04 00:00:00",
    "user_id": 1,
    "canonical": "hello-world",
    "title": "Hello World!",
    "deleted": null,
    "updated": "2019-05-25 00:36:52",
    "id": 1,
  },
  "saveContentBatch": [Function],
  "saveContentBatchDebounce": [Function],
}
`;

exports[`UpdateManager stageNodeDelete method 1`] = `
UpdateManager {
  "commitTimeoutId": undefined,
  "lastUndoHistoryPush": undefined,
  "post": Immutable.Map {
    "abstract": "Creating a simple React ( + Babel + Webpack + yarn) starter project is actually proving to be difficult starting from the documentation.",
    "created": "2019-05-18 17:28:55",
    "published": "2019-01-04 00:00:00",
    "user_id": 1,
    "canonical": "hello-world",
    "title": "Hello World!",
    "deleted": null,
    "updated": "2019-05-25 00:36:52",
    "id": 1,
  },
  "saveContentBatch": [Function],
  "saveContentBatchDebounce": [Function],
}
`;

exports[`UpdateManager stageNodeUpdate method 1`] = `
UpdateManager {
  "commitTimeoutId": undefined,
  "lastUndoHistoryPush": undefined,
  "post": Immutable.Map {
    "abstract": "Creating a simple React ( + Babel + Webpack + yarn) starter project is actually proving to be difficult starting from the documentation.",
    "created": "2019-05-18 17:28:55",
    "published": "2019-01-04 00:00:00",
    "user_id": 1,
    "canonical": "hello-world",
    "title": "Hello World!",
    "deleted": null,
    "updated": "2019-05-25 00:36:52",
    "id": 1,
  },
  "saveContentBatch": [Function],
  "saveContentBatchDebounce": [Function],
}
`;

exports[`UpdateManager updates method 1`] = `
Array [
  Array [
    "foo",
    Object {
      "action": "update",
      "node": Object {
        "id": "foo",
      },
      "post_id": 1,
    },
  ],
  Array [
    "qux",
    Object {
      "action": "update",
      "node": Object {
        "id": "qux",
      },
      "post_id": 1,
    },
  ],
  Array [
    "bar",
    Object {
      "action": "delete",
      "post_id": 1,
    },
  ],
]
`;
